File: RaceControlAndroid/app/src/main/java/com/owlmedia/racecontrol/feature/standings/WdcCalculatorView.kt
package com.owlmedia.racecontrol.feature.standings

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.EmojiEvents
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.owlmedia.racecontrol.R
import com.owlmedia.racecontrol.core.design.Dimens
import com.owlmedia.racecontrol.core.design.RcShapes
import com.owlmedia.racecontrol.core.design.RcTheme
import com.owlmedia.racecontrol.core.design.tabular
import com.owlmedia.racecontrol.core.ui.EmptyState
import com.owlmedia.racecontrol.core.ui.LoadableContent
import com.owlmedia.racecontrol.core.ui.TeamLogo
import com.owlmedia.racecontrol.data.remote.dto.WdcDriverDto
import com.owlmedia.racecontrol.feature.AppState

/**
 * Who can still mathematically win the drivers' championship.
 *
 * Mirrors the web/iOS "WDC calculator": for every driver, the ceiling is
 * their current points plus every remaining point on offer (race + sprint),
 * assuming they win everything left and the leader scores nothing else. That
 * is a best case, not a forecast. The caption under the header exists
 * because a user found the raw numbers confusing without it.
 */
@Composable
fun WdcCalculatorView(
    viewModel: StandingsViewModel,
    appState: AppState,
    onOpenDriver: (String) -> Unit = {},
) {
    val state by viewModel.wdc.collectAsStateWithLifecycle()

    LoadableContent(
        state = state,
        onRetry = { viewModel.load(appState.selectedYear, StandingsMode.WDC, force = true) },
    ) { data ->
        if (data.drivers.isEmpty()) {
            EmptyState(
                icon = Icons.Filled.EmojiEvents,
                title = stringResource(R.string.no_standings_title),
                message = stringResource(R.string.no_standings_message, appState.selectedYear.toString()),
            )
            return@LoadableContent
        }

        val canWinCount = data.drivers.count { it.canWin }

        LazyColumn(
            contentPadding = PaddingValues(Dimens.MD),
            verticalArrangement = Arrangement.spacedBy(Dimens.SM),
        ) {
            item(key = "wdc-header") {
                WdcHeader(
                    decided = data.decided,
                    roundsRemaining = data.roundsRemaining,
                    maxRemainingPoints = data.maxRemainingPoints,
                    canWinCount = canWinCount,
                )
            }
            items(
                items = data.drivers,
                key = { it.driverId ?: it.fullName },
                contentType = { "wdc-driver" },
            ) { driver ->
                WdcDriverRow(driver = driver, onOpenDriver = onOpenDriver)
            }
        }
    }
}

@Composable
private fun WdcHeader(
    decided: Boolean,
    roundsRemaining: Int,
    maxRemainingPoints: Int,
    canWinCount: Int,
) {
    val statusText = when {
        decided && roundsRemaining == 0 -> "Season complete: the title is decided."
        decided -> "Mathematically settled: only the leader can still win."
        else -> {
            val roundsWord = if (roundsRemaining == 1) "round" else "rounds"
            val driversWord = if (canWinCount == 1) "driver" else "drivers"
            "$roundsRemaining $roundsWord left · up to $maxRemainingPoints points still on offer · " +
                "$canWinCount $driversWord can still win"
        }
    }

    Column(modifier = Modifier.padding(bottom = Dimens.XS)) {
        Text(
            text = statusText,
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.SemiBold,
            color = RcTheme.colors.textPrimary,
        )
        Spacer(Modifier.height(2.dp))
        Text(
            text = "\"Can win\" is the best-case ceiling (winning every remaining session while the " +
                "leader scores nothing else), not a realistic forecast.",
            style = MaterialTheme.typography.bodySmall,
            color = RcTheme.colors.textTertiary,
        )
    }
}

@Composable
private fun WdcDriverRow(
    driver: WdcDriverDto,
    onOpenDriver: (String) -> Unit,
) {
    val behindLabel = if (driver.pointsBehindLeader == 0.0) "Leader" else "-${formatPoints(driver.pointsBehindLeader)}"

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RcShapes.Medium)
            .background(MaterialTheme.colorScheme.surface)
            .padding(Dimens.SM)
            .semantics(mergeDescendants = true) {
                contentDescription = buildString {
                    append("Position ${driver.position ?: 0}, ${driver.fullName}")
                    if (!driver.teamName.isNullOrBlank()) append(", ${driver.teamName}")
                    append(", ${formatPoints(driver.points)} points")
                    append(", ${if (driver.pointsBehindLeader == 0.0) "leader" else "${formatPoints(driver.pointsBehindLeader)} behind leader"}")
                    append(", max possible ${formatPoints(driver.maxPoints)} points")
                    append(if (driver.canWin) ", can still win" else ", cannot win")
                }
            },
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                text = (driver.position ?: 0).toString(),
                style = MaterialTheme.typography.titleMedium.tabular(),
                fontWeight = FontWeight.Bold,
                color = if ((driver.position ?: 0) <= 3) RcTheme.colors.gold else RcTheme.colors.textSecondary,
                modifier = Modifier.width(28.dp),
            )
            TeamLogo(url = driver.teamLogoUrl, modifier = Modifier.padding(end = Dimens.SM))
            Column(Modifier.weight(1f)) {
                Text(
                    text = driver.fullName,
                    style = MaterialTheme.typography.bodyLarge,
                    fontWeight = FontWeight.SemiBold,
                    color = if (driver.driverId != null) RcTheme.colors.info else RcTheme.colors.textPrimary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.clickable(enabled = driver.driverId != null) {
                        onOpenDriver(driver.driverId ?: return@clickable)
                    },
                )
                if (!driver.teamName.isNullOrBlank()) {
                    Text(
                        text = driver.teamName,
                        style = MaterialTheme.typography.bodySmall,
                        color = RcTheme.colors.textSecondary,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
            CanWinPill(canWin = driver.canWin)
        }
        Spacer(Modifier.height(6.dp))
        Row(horizontalArrangement = Arrangement.spacedBy(Dimens.MD)) {
            WdcStat(label = "Points", value = formatPoints(driver.points))
            WdcStat(label = "Behind", value = behindLabel)
            WdcStat(label = "Max possible", value = formatPoints(driver.maxPoints))
        }
    }
}

@Composable
private fun WdcStat(label: String, value: String) {
    Column {
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = RcTheme.colors.textTertiary,
        )
        Text(
            text = value,
            style = MaterialTheme.typography.labelLarge.tabular(),
            fontWeight = FontWeight.Bold,
            color = RcTheme.colors.textPrimary,
        )
    }
}

@Composable
private fun CanWinPill(canWin: Boolean) {
    val background = if (canWin) RcTheme.colors.positive.copy(alpha = 0.2f) else RcTheme.colors.surfaceElevated
    val textColor = if (canWin) RcTheme.colors.positive else RcTheme.colors.textTertiary
    val label = if (canWin) "Can win" else "Can't win"

    Box(
        modifier = Modifier
            .clip(RcShapes.Small)
            .background(background)
            .padding(horizontal = Dimens.SM, vertical = 4.dp),
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            fontWeight = FontWeight.Bold,
            color = textColor,
        )
    }
}

/** Points without a trailing ".0": F1 points are integers apart from half-points races. */
private fun formatPoints(value: Double): String =
    if (value % 1.0 == 0.0) value.toInt().toString() else value.toString()

File: RaceControlApp/RaceControl/Features/Standings/WdcCalculatorView.swift
import SwiftUI

/// Drivers' championship "who can still win" calculator: for each driver still
/// in contention, shows their mathematical ceiling if they won every remaining
/// point-scoring opportunity while the current leader scored nothing else.
struct WdcCalculatorView: View {
    let year: Int
    @StateObject private var vm = WdcCalculatorViewModel()

    var body: some View {
        LoadableView(state: vm.state) {
            await vm.load(year: year)
        } content: { data in
            content(data)
        }
        .background(Theme.Palette.background)
        .task(id: year) { await vm.load(year: year) }
    }

    private func content(_ data: WdcCalculator) -> some View {
        ScrollView {
            VStack(spacing: Theme.Space.sm) {
                statusHeader(data)
                pointsBreakdown
                LazyVStack(spacing: Theme.Space.sm) {
                    ForEach(data.drivers) { driver in
                        WdcDriverRow(driver: driver, year: year, driversById: vm.driversById)
                    }
                }
            }
            .padding(Theme.Space.md)
        }
    }

    @ViewBuilder
    private func statusHeader(_ data: WdcCalculator) -> some View {
        let canWinCount = data.drivers.filter(\.canWin).count
        VStack(alignment: .leading, spacing: 4) {
            Text(statusLine(data, canWinCount: canWinCount))
                .font(.subheadline.weight(.semibold))
                .foregroundStyle(Theme.Palette.textPrimary)
            Text("\u{201C}Can win\u{201D} is the best-case ceiling (winning every remaining session while the leader scores nothing else), not a realistic forecast.")
                .font(.caption)
                .foregroundStyle(Theme.Palette.textSecondary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Theme.Space.md)
        .background(Theme.Palette.surface, in: RoundedRectangle(cornerRadius: Theme.Radius.md))
        .overlay(RoundedRectangle(cornerRadius: Theme.Radius.md).stroke(Theme.Palette.stroke, lineWidth: 1))
    }

    private func statusLine(_ data: WdcCalculator, canWinCount: Int) -> String {
        if data.decided && data.roundsRemaining == 0 {
            return "Season complete: the title is decided."
        }
        if data.decided {
            return "Mathematically settled: only the leader can still win."
        }
        let roundWord = data.roundsRemaining == 1 ? "round" : "rounds"
        let driverWord = canWinCount == 1 ? "driver" : "drivers"
        return "\(data.roundsRemaining) \(roundWord) left · up to \(data.maxRemainingPoints) points still on offer · \(canWinCount) \(driverWord) can still win"
    }

    private var pointsBreakdown: some View {
        DisclosureGroup {
            VStack(alignment: .leading, spacing: Theme.Space.xs) {
                Text("Race (P1–P10): 25, 18, 15, 12, 10, 8, 6, 4, 2, 1")
                Text("Sprint (P1–P8): 8, 7, 6, 5, 4, 3, 2, 1")
                Text("Fastest lap bonus: +1 (top 10 finishers, race only)")
            }
            .font(.caption)
            .foregroundStyle(Theme.Palette.textSecondary)
            .padding(.top, Theme.Space.sm)
            .frame(maxWidth: .infinity, alignment: .leading)
        } label: {
            Text("POINTS BREAKDOWN")
                .font(.caption.weight(.bold)).tracking(1)
                .foregroundStyle(Theme.Palette.textSecondary)
        }
        .padding(Theme.Space.md)
        .background(Theme.Palette.surface, in: RoundedRectangle(cornerRadius: Theme.Radius.md))
        .overlay(RoundedRectangle(cornerRadius: Theme.Radius.md).stroke(Theme.Palette.stroke, lineWidth: 1))
    }
}

private struct WdcDriverRow: View {
    let driver: WdcDriverEntry
    let year: Int
    let driversById: [String: Driver]

    private var matchedDriver: Driver? {
        guard let id = driver.driverId else { return nil }
        return driversById[id]
    }

    var body: some View {
        HStack(spacing: Theme.Space.md) {
            PositionBadge(text: String(driver.position ?? 0), highlight: (driver.position ?? 0) <= 3)
            TeamAccentBar(color: .team(driver.teamColor))
            TeamLogoView(url: driver.teamLogoUrl, size: 28)

            VStack(alignment: .leading, spacing: 2) {
                nameLabel
                Text(driver.teamName ?? "")
                    .font(.caption)
                    .foregroundStyle(Theme.Palette.textSecondary)
                    .lineLimit(1)
            }

            Spacer()

            VStack(alignment: .trailing, spacing: 2) {
                HStack(alignment: .firstTextBaseline, spacing: Theme.Space.sm) {
                    VStack(alignment: .trailing, spacing: 0) {
                        Text(numberLabel(driver.points))
                            .font(.system(.subheadline, design: .rounded).weight(.bold))
                            .monospacedDigit()
                            .foregroundStyle(Theme.Palette.textPrimary)
                        Text("PTS").font(.caption2).foregroundStyle(Theme.Palette.textTertiary)
                    }
                    VStack(alignment: .trailing, spacing: 0) {
                        Text(behindLabel)
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(driver.pointsBehindLeader == 0 ? Theme.Palette.positive : Theme.Palette.textSecondary)
                        Text("MAX \(numberLabel(driver.maxPoints))")
                            .font(.caption2)
                            .foregroundStyle(Theme.Palette.textTertiary)
                    }
                }
                canWinPill
            }
        }
        .padding(Theme.Space.md)
        .background(Theme.Palette.surface, in: RoundedRectangle(cornerRadius: Theme.Radius.md))
        .overlay(RoundedRectangle(cornerRadius: Theme.Radius.md).stroke(Theme.Palette.stroke, lineWidth: 1))
    }

    @ViewBuilder
    private var nameLabel: some View {
        if let matchedDriver {
            NavigationLink {
                DriverDetailView(year: year, driver: matchedDriver)
            } label: {
                Text(driver.fullName)
                    .font(.headline)
                    .foregroundStyle(Theme.Palette.textPrimary)
                    .lineLimit(1)
            }
        } else {
            Text(driver.fullName)
                .font(.headline)
                .foregroundStyle(Theme.Palette.textPrimary)
                .lineLimit(1)
        }
    }

    private var behindLabel: String {
        driver.pointsBehindLeader == 0 ? "Leader" : "-\(numberLabel(driver.pointsBehindLeader))"
    }

    private var canWinPill: some View {
        Text(driver.canWin ? "Can win" : "Can't win")
            .font(.caption2.weight(.bold))
            .foregroundStyle(driver.canWin ? Theme.Palette.positive : Theme.Palette.textTertiary)
            .padding(.horizontal, 8)
            .padding(.vertical, 3)
            .background(
                (driver.canWin ? Theme.Palette.positive : Theme.Palette.textTertiary).opacity(0.15),
                in: Capsule()
            )
    }

    private func numberLabel(_ value: Double) -> String {
        value.truncatingRemainder(dividingBy: 1) == 0 ? String(Int(value)) : String(format: "%.1f", value)
    }
}

@MainActor
final class WdcCalculatorViewModel: ObservableObject {
    @Published var state: Loadable<WdcCalculator> = .idle
    @Published var driversById: [String: Driver] = [:]
    private var loadedYear: Int?

    func load(year: Int) async {
        if loadedYear == year, case .loaded = state { return }
        state = .loading
        do {
            state = .loaded(try await APIClient.shared.wdcCalculator(year: year))
            loadedYear = year
        } catch {
            state = .failed((error as? APIError)?.errorDescription ?? error.localizedDescription)
        }
        // Secondary lookup for driver-detail navigation; failure shouldn't block the screen.
        if let drivers = try? await APIClient.shared.drivers(year: year) {
            driversById = Dictionary(uniqueKeysWithValues: drivers.map { ($0.driverId, $0) })
        }
    }
}
